synapse/http/proxyagent.py

```diff
@@ -208,30 +210,46 @@ def request(self, method, uri, headers=None, bodyProducer=None):
         )
 
 
-def _http_proxy_endpoint(proxy, reactor, **kwargs):
+def _http_proxy_endpoint(
+    proxy: Optional[bytes],
+    reactor: IReactorCore,
+    tls_options_factory: Optional[IPolicyForHTTPS],
+    **kwargs
+):
     """Parses an http proxy setting and returns an endpoint for the proxy
 
     Args:
         proxy (bytes|None):  the proxy setting
         reactor: reactor to be used to connect to the proxy
+        tls_options_factory: the TLS options to use when connecting through a https proxy
         kwargs: other args to be passed to HostnameEndpoint
 
     Returns:
         interfaces.IStreamClientEndpoint|None: endpoint to use to connect to the proxy,
             or None
+
+    Raises: ValueError if given a proxy with a scheme we don't support.
     """
     if proxy is None:
         return None
 
-    # currently we only support hostname:port. Some apps also support
-    # protocol://<host>[:port], which allows a way of requiring a TLS connection to the
-    # proxy.
+    # Note: we can't use urlsplit/urlparse as that is completely broken for things without a scheme://
+    schemehost, port = parse_host_port(proxy, default_port=1080)
+    scheme, host = parse_scheme_host(schemehost, default_scheme=b"http")
```

Comment:
There was a previous conversation about this (https://github.com/matrix-org/synapse/pull/9119#discussion_r562855935) but I want to revisit it.

I did some quick testing and using `urlparse` here seems to work fine for everything I've tried:

```pycon
>>> from urllib.parse import urlparse
>>> urlparse("proxy.foo", scheme="http")
ParseResult(scheme='http', netloc='', path='proxy.foo', params='', query='', fragment='')
>>> urlparse("proxy.foo:9000", scheme="http")
ParseResult(scheme='http', netloc='', path='proxy.foo:9000', params='', query='', fragment='')
>>> urlparse("http://proxy.foo:9000", scheme="http")
ParseResult(scheme='http', netloc='proxy.foo:9000', path='', params='', query='', fragment='')
>>> urlparse("https://proxy.foo:9000", scheme="http")
ParseResult(scheme='https', netloc='proxy.foo:9000', path='', params='', query='', fragment='')
>>> urlparse("https://proxy.foo", scheme="http")
ParseResult(scheme='https', netloc='proxy.foo', path='', params='', query='', fragment='')
```

So I really think we should be using `urlparse` here. Can you provide a test-case in which it doesn't work?

Reply:
uhm... your second example fails horribly for me:

```
$ python
Python 3.9.2 (default, Feb 20 2021, 18:40:11) 
[GCC 10.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from urllib.parse import urlparse
>>> urlparse("proxy.foo:9000", scheme="http")
ParseResult(scheme='proxy.foo', netloc='', path='9000', params='', query='', fragment='')
>>> 
```

Reply:
the first one results in:
```
urlparse("proxy.foo", scheme="http")
ParseResult(scheme='http', netloc='', path='proxy.foo', params='', query='', fragment='')
```

Reply:
(Solved all other comments, which leaves only this open)

This seems to have been changed for the (far) worse in python 3.9... which makes me not want to rely on this function ever again, really. I'd just leave that as it is now?

Reply:
This is very confusing, but it seems that there was a change in python/cpython#661 to make `urlparse` match the RFC closer that resulted in backwards incompatible behavior. (This *did* get backported to 3.7.5, but then got reverted for 3.7.7, similarly for 3.8.)

It does appear in the [Python 3.9 changelog](https://docs.python.org/3.9/whatsnew/changelog.html):

> [bpo-27657](https://bugs.python.org/issue27657): Fix urllib.parse.urlparse() with numeric paths. A string like “path:80” is no longer parsed as a path but as a scheme (“path”) and a path (“80”).

There's a whole slew of other bugs linked from the above...anyway, it seems we'll need to handle this in a custom matter. I'll look more closely at that code now!